sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -169,11 +179,30 @@ class StreamExecution(
    * least the given `Offset`. This method is indented for use primarily when writing tests.
    */
   def awaitOffset(source: Source, newOffset: Offset): Unit = {
-    while (!currentOffsets.contains(source) || currentOffsets(source) < newOffset) {
+    def notDone = synchronized {
+      !currentOffsets.contains(source) || currentOffsets(source) < newOffset
+    }
+
+    while (notDone) {
       logInfo(s"Waiting until $newOffset at $source")
-      synchronized { wait() }
+      awaitBatchLock.synchronized { awaitBatchLock.wait(100) }
     }
     logDebug(s"Unblocked at $newOffset for $source")
   }
+
+  /** Clears the indicator that a batch has completed.  Used for testing. */
+  override def clearBatchMarker(): Unit = {
+    batchRun = false
+  }
+
+  /**
+   * Awaits the completion of at least one streaming batch. Must be called after `clearBatchMarker`
+   * to gurantee that a new batch has been processed.
+   */
+  override def awaitNextBatch(): Unit = {
+    while (!batchRun) {
+      awaitBatchLock.synchronized { awaitBatchLock.wait(100) }
```

Comment:
Thats a really good question.  When I had it as `wait()`, it was hanging non-deterministically.  I think its okay to spin occasionally?